test/framework/clusterresourceset_helpers.go

```diff
@@ -150,7 +151,12 @@ func WaitForClusterResourceSetToApplyResources(ctx context.Context, input WaitFo
 				continue
 			}
 
-			if len(binding.Spec.Bindings) == 0 || !binding.Spec.Bindings[0].IsApplied(resource) {
+			// Check all bindings to see if the resource is applied. If multiple ClusterResourceSets match a cluster, the
+			// ClusterResourceSetBinding will have multiple bindings and so every binding must be checked.
+			applied := slices.ContainsFunc(binding.Spec.Bindings, func(b *addonsv1.ResourceSetBinding) bool {
+				return b.IsApplied(resource)
+			})
+			if !applied {
 				return false
 			}
```

Comment:
```suggestion
			for _, binding := range binding.Spec.Bindings {
				if !binding.IsApplied(resource) {
					return false
				}
			}
```

I think this is a more readable way to check this :-)

Thanks for finding this and opening a PR!

Reply:
I'd have expected that we only want to return true if all bindings of a crs are applied.
If one is not applied we return false to let `Eventually` try it later again?

Reply:
When a `ClusterResourceSet` matches a cluster, a `ClusterResourceSetBinding` is created for that specific cluster. If multiple `ClusterResourceSets` match a cluster then the `ClusterResourceSetBinding` is updated to include all the matching `ClusterResourceSets` and the state of the resources they reference (i.e. is each of them applied).

This function, `WaitForClusterResourceSetToApplyResources`, is checking that the resources referenced from the single specified `ClusterResourceSet` are all applied. There are potentially other `ResourceBindings` that are not related to the `ClusterResourceSet` that this function is waiting on, hence we need to check all `ResourceBinding§ to check the resource is actually applied.

I've talked myself into changing this implementation to first find the relevant `ResourceBinding` from `spec.bindings` by matching the `ClusterResourceSet` name, and then calling what you suggest above to return false quickly if the resource has not been applied.

Does that make sense?

Reply:
Ah now I got it. Missed the point that we just want to check a single CRS 👍 thanks for helping my brain out 🚀

Reply:
Thanks for helping my brain out too! 🙏